### DISCUSSION TITLE
@lit-labs/react properties not available when connectedCallback is called

### DISCUSSION THREAD

**Discussion by OvidijusParsiunas** [lit/lit#3649] [Q&A] [answered]
Hi, I am using the `@lit-labs/react` package to render my custom element in jsx files. Everything works fine except of one problem, the properties that I set are not available when the `connectedCallback` lifecycle method is called and are only available at a later stage.

In the jsx file if I have the following:
```
<ComponentWrapper greeting="hello"></ComponentWrapper>
```

Custom element file:
```
connectedCallback() {
    super.connectedCallback();
    console.log(this.greeting); // undefined printed :(
    setTimeout(() => {
      console.log(this.greeting); // 'hello' printed
    });
  }
```

[LIVE EXAMPLE](https://stackblitz.com/edit/react-szfq66?file=src/simple-greeting-component.js)
(If you see an error on the screen - make sure to hit the refresh button on top right)

The use of `setTimeout` solves my problem and thankfully the code inside it runs before the `render` lifecycle method is called. However this seems like a hack around an actual issue and I was wondering if there is a more straightforward way to use properties before `render` is called?

Thankyou.

**Comment by augustjk** [comment, marked as answer]
Edit: For OP's use case needing to pre-process props before `render()`, the solution is to override `update()` and do the work before `super.update()`.

Original response below:

This is due to the property setting happening on the `componentDidMount()` lifecycle for React which is happens after the custom element is added to the DOM and is connected.

Is it possible to move the code that needs access to the property into the [`firstUpdated()`](https://lit.dev/docs/components/lifecycle/#firstupdated) lifecycle (if the custom element is authored with Lit)?

**Reply by OvidijusParsiunas** [reply]
`firstUpdated()` runs after `render` which is not ideal :/

**Reply by OvidijusParsiunas** [reply]
My goal is to pre-process the passed in properties and create elements with events depending on those properties before `render` is called.

**Reply by OvidijusParsiunas** [reply]
The use of `setTimeout` does work, but I am leaving the decision of when it gets triggered to the event loop, which is outside of my control and may be prone to issues if certain frameworks/browsers work in a different way.

**Reply by justinfagnani** [reply]
sounds like a job for overriding `update()` and doing work before `super.update()`

**Reply by OvidijusParsiunas** [reply]
This works perfectly, thankyou very much! Not sure if you want me to mark the op of this thread as the answer or you can create a new one and I can mark that.
